### PR TITLE
Add GitHub Actions configuration file

### PR DESCRIPTION
**Pull request by smorimoto**
This PR adds a configuration file for GitHub Actions.

There are jobs that don't build and/or test for a variety of reasons, but they are all commented out. See the diagram below for what to do on which version and platform.

| [VERSION] - [PLATFORM]          | Build              | Test               |
| ------------------------------- | ------------------ | ------------------ |
| 4.09.0 - macOS, Ubuntu, Windows | :heavy_check_mark: | :heavy_check_mark: |
| 4.08.1 - macOS, Ubuntu, Windows | :heavy_check_mark: | :heavy_check_mark: |
| 4.07.1 - macOS, Ubuntu, Windows | :heavy_check_mark: | :x:                |
| 4.02.3 - macOS, Ubuntu          | :heavy_check_mark: | :x:                |
| 4.02.3 - Windows                | :x:                | :x:                |

### CODE REVIEW

**Comment by jeremiedimino** [comment]
Thanks for playing with github actions for dune :)

Would this replace Travis and AppVeyor?

**Comment by imbsky** [comment]
Yes, that's right. and this works without complex shell scripts.

**Comment by jeremiedimino** [comment]
That's nice.

/cc our CI experts: @avsm @dra27 @CraigFe

**Comment by imbsky** [comment]
By the way, you can check here that it works correctly.
https://github.com/imbsky/dune/actions/runs/96236696

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      - run: opam pin add dune-action-plugin.dev . --no-action
+      - run: opam pin add dune-build-info.dev . --no-action
+      - run: opam pin add dune-configurator.dev . --no-action
+      - run: opam pin add dune-glob.dev . --no-action
+      - run: opam pin add dune-private-libs.dev . --no-action
+      - run: opam pin add dune.dev . --no-action
+
+      # Our current setup for 4.02 doesn’t use opam, So skip the next three
+      # steps on 4.02.3
+      - run: opam depext dune-action-plugin dune-build-info dune-configurator dune-glob dune-private-libs dune --yes --with-test --with-doc
```
Comment (rgrinberg) [inline .github/workflows/workflow.yml:67]: Why do we need this step? We don't use any depexts in dune, and we aren't planning to.

File: .github/workflows/workflow.yml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        if: matrix.ocaml-version != '4.02.3'
+
+      - run: opam install . --deps-only --with-test --with-doc
+        if: matrix.ocaml-version != '4.02.3'
+
+      - run: opam install coq mdx menhir ocaml-migrate-parsetree ocamlfind odoc ppx_expect ppxlib result utop
```
Comment (rgrinberg) [inline .github/workflows/workflow.yml:73]: Instead of this, would it be possible to run a step that installs the `DEV_DEPS` in the makefile?

I'm aware there might not be a convenient makefile target at the moment, but it could be a nice thing to add.

File: .github/workflows/workflow.yml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+      - run: opam exec -- make dune.exe
+
+      # We don't run tests on all versions of the Windows environment and on
+      # 4.02.3 and 4.07.1 in other environments
+      - run: opam exec -- make test
+        if: runner.os != 'Windows' && matrix.ocaml-version != '4.02.3' && matrix.ocaml-version != '4.07.1'
+
+      - run: opam exec -- make test-coq
+        if: runner.os != 'Windows' && matrix.ocaml-version != '4.02.3' && matrix.ocaml-version != '4.07.1'
```
Comment (rgrinberg) [inline .github/workflows/workflow.yml:84]: Is it possible to extract this condition into a variable? I'm not familiar with the github actions, but it would be really nice not to repeat this logic.

File: .github/workflows/workflow.yml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      # Our current setup for 4.02 doesn’t use opam, So skip the next three
+      # steps on 4.02.3
+      - run: opam depext dune-action-plugin dune-build-info dune-configurator dune-glob dune-private-libs dune --yes --with-test --with-doc
+        if: matrix.ocaml-version != '4.02.3'
+
+      - run: opam install . --deps-only --with-test --with-doc
+        if: matrix.ocaml-version != '4.02.3'
+
+      - run: opam install coq mdx menhir ocaml-migrate-parsetree ocamlfind odoc ppx_expect ppxlib result utop
+        if: matrix.ocaml-version != '4.02.3'
```
Comment (dra27) [inline .github/workflows/workflow.yml:74]: These dependencies are only needed for the tests, so you should be able to speed up Windows considerably by having the same test condition as below?

**Comment by dra27** [comment]
Nice! No reason for a short period not to run it alongside AppVeyor/Travis and then retire them a few weeks after merging?

**Comment by dra27** [comment]
> Yes, that's right. and this works without complex shell scripts.

I'm not against it at all, but this is because the complex logic is offloaded to another action!

Another important caveat at the moment is run time - part of the complexity (especially on Travis) is managing the caching. The action is presently taking half an hour to run where the AppVeyor takes 90 seconds and Travis around 17 minutes.

**Comment by jeremiedimino** [comment]
Run time is definitely a concern. If there are other benefits to switching to github actions, we can accept a small slowdown, but going from 17 minutes to half an hour is too much.

**Comment by rgrinberg** [comment]
The run time increase is worrying and bit unexpected. When we switched lsp to github actions from travis, our run times improved significantly.

**Comment by jeremiedimino** [comment]
Maybe it's because of the caching @dra27 described. Do you do anything clever in lsp regarding opam caching?

**Comment by rgrinberg** [comment]
I don't think we have any special caching (@giltho can confirm). But we also don't install nearly as much dependencies for tests.

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (imbsky) [reply .github/workflows/workflow.yml]: I removed them.

File: .github/workflows/workflow.yml (hunk 2 above)
Comment (imbsky) [reply .github/workflows/workflow.yml]: Changed.

File: .github/workflows/workflow.yml (hunk 3 above)
Comment (imbsky) [reply .github/workflows/workflow.yml]: I tried to do it, but I don't know if it works correctly. Let's look at the CI results.

File: .github/workflows/workflow.yml (hunk 4 above)
Comment (imbsky) [reply .github/workflows/workflow.yml]: I did it with other changes.

**Comment by imbsky** [comment]
> > Yes, that's right. and this works without complex shell scripts.
>
> I'm not against it at all, but this is because the complex logic is offloaded to another action!
>
> Another important caveat at the moment is run time - part of the complexity (especially on Travis) is managing the caching. The action is presently taking half an hour to run where the AppVeyor takes 90 seconds and Travis around 17 minutes.

Yeah, you're absolutely right! My (and our?) goal is to push the many complex tasks that are going on in each repository to a particular Action.

**Comment by imbsky** [comment]
For example, I've been working for about six months recently to speed up `actions/cache`, mainly for OCaml projects, and it's about 4~5 times faster as a result. This is probably good for caching dependencies, etc.

I also plan to rewrite the `avsm/setup-ocaml` to cache the contents of the opam directory so that we can set up our OCaml environment faster.

**Comment by giltho** [comment]
Yep confirming that we have no caching at all in ocaml-lsp's CI
As far as I know there is no good caching story yet for opam + GH actions workflow yet. But @imbsky is a maintainer of setup-ocaml and is working on caching there as said earlier.

The main advantage of moving to gha is indeed to have all the complexity encapsulated in an action instead of having custom scripts. Dune is a major engine that moves the community forward and it would be very nice for it to use gha so that the action can improve, and therefore improve for the whole community :)

That being said, it would probably be better to wait imo for the caching to work before merging this, or at least not remove the rest of the CI and make the gha checks optional for merging.

**Comment by imbsky** [comment]
I'm sorry for committing so much. I was caught in some traps. I think I was able to reflect all　suggestions in code review.

File: .github/workflows/workflow.yml (hunk 3 above)
Comment (imbsky) [reply .github/workflows/workflow.yml]: I've dealt to this, but it forces a less intuitive writing style. and we can't access the `runner.os` context outside of the steps field, so the env field can be very verbose, for example when testing on multiple Windows versions.

Comment (imbsky) [reply .github/workflows/workflow.yml]: This writing style reminds me of this problem. https://ariya.io/2011/08/hall-of-api-shame-boolean-trap

**Comment by jeremiedimino** [comment]
No worries, we'll probably just squash everything before merging. Committing a lot during development is generally good practice :)

Otherwise, I'm personally happy to move to github actions. It's nice to have all the platforms tested in one place. But the run time is definitely a concern, and we should indeed wait until it is close enough to the current run time we get with the combo Travis/AppVeyor.

**Comment by imbsky** [comment]
Thank you for your kind words. Yes, it makes sense. I will mark this PR as draft until their speed improves!

**Comment by rgrinberg** [comment]
I looked at the CI configuration and it doesn't seem like the run times are comparable. Here are some differences:

* We are running the job on 3 versions of OCaml on windows, while the current appveyor runs it only one version of OCaml.

* We are using opam for the windows jobs. In our current setup, we don't use windows.

* We are running the tests only on one version of OCaml and a single OS (Ubuntu). With actions, we are running the tests on 4.07-4.09 and on Mac and Linux.

* We are testing that 4.02, 4.07, 4.08, 4.09 build dune on ALL platforms rather than just one.

In summary, no wonder github actions seems a lot slower -- we are testing quite a bit more configurations.  I would suggest the following changes:

* Run the 4.02 build only on a single platform (Linux)
* Run the test suite only on the latest version of OCaml on Mac/Linux.
* Add a CI configuration (build only) that does not use opam. It's important for dune to be usable without opam, so I don't think we should drop this configuration.

**Comment by rgrinberg** [comment]
Once https://github.com/ocaml/dune/pull/3466 is merged, we can drop installing gstat with homebrew by the way.

**Comment by imbsky** [comment]
Running jobs on all platforms and versions are done in parallel, so I don't think it will speed up CI.
The most time-consuming part is setting up an OCaml environment and installing dependencies. So, in Windows, for example, only builds and doesn't install any extra dependencies, the job will finish in about 5 minutes.

**Comment by imbsky** [comment]
If we want to speed everything up, we need to set up the OCaml environment within seconds to a minute. Because GitHub Actions has a cache limit for users, at least for now.

> A repository can have up to 5GB of caches. Once the 5GB limit is reached, older caches will be evicted based on when the cache was last accessed. Caches that are not accessed within the last week will also be evicted.

But, on the Action side is unlimited. So I think we can have the pre-built switch on the setup-ocaml side. And that's what I'm working on now.

**Comment by dra27** [comment]
@rgrinberg - the important figure, which his comparable is "How long does it all take"... reducing the testing matrix is a fix for that, being able to parallelise is another! 🙂

**Comment by voodoos** [comment]
I am sorry if my remark sounds naive, but I have a very simple project in which I used [`actions/cache`](https://github.com/actions/cache) to chache the entire `.opam` folders and in my setup it works quite well:

```
steps:
    - uses: actions/checkout@v2
    - name: Cache
      uses: actions/cache@v1.1.2
      with:
        # A directory to store and save the cache
        path: ~/.opam
        # An explicit key for restoring and saving the cache
        key: ${{ runner.os }}-build
    - name: Set up OCaml
      uses: avsm/setup-ocaml@v1.0.1
      with:
        # Version of the OCaml compiler to initialise
        ocaml-version: 4.09.1
    - ...
```

Is this design fundamentally flawed for more complex settings like Dune's CI ?

**Comment by imbsky** [comment]
It works for small projects, but the problem is projects like this. The cache limit is only 5 GB and will likely result in cache thrashing which will not get the expected results.
https://github.com/actions/cache#cache-limits

**Comment by voodoos** [comment]
> 
> 
> It works for small projects, but the problem is projects like this. The cache limit is only 5 GB and will likely result in cache thrashing which will not get the expected results.
> https://github.com/actions/cache#cache-limits

Ok I see, thanks for you answer !

(Apparently I accidentality deleted my question: I am using `action/cache` to cache the whole `.opam` folder on a small project of mine and it works great, I was wondering why it wouldn't scale to Dune CI.)

**Comment by rgrinberg** [comment]
@imbsky are there any updates on the caching issue? we've discussed this PR at the recent dune meeting and there was renewed interest in it.

**Comment by imbsky** [comment]
I started a personal experiment project like this. Hopefully this will reduce the compilation time of the compiler itself. I want to work on this project as soon as I get home after this week's exams and so on.
https://github.com/actions-ml

**Comment by smorimoto** [comment]
To make it faster, I want to work more closely with the GitHub team rather than doing it in my own way, and while the project above is a bit stalled, things are moving forward.

**Comment by jeremiedimino** [comment]
That's good to hear :) We look forward to a simpler and more centralised CI, the current one based on Travis and AppVeyor is regularly causing us pain.

**Comment by rgrinberg** [comment]
Replaced by #3712.
